File: src/mini_app/serializers/analytics.py
from __future__ import annotations

from typing import Any

from application.services.stats import HelpdeskAnalyticsSnapshot


def serialize_analytics_snapshot(snapshot: HelpdeskAnalyticsSnapshot) -> dict[str, Any]:
    return {
        "window": snapshot.window.value,
        "total_open_tickets": snapshot.total_open_tickets,
        "queued_tickets_count": snapshot.queued_tickets_count,
        "assigned_tickets_count": snapshot.assigned_tickets_count,
        "escalated_tickets_count": snapshot.escalated_tickets_count,
        "closed_tickets_count": snapshot.closed_tickets_count,
        "period_created_tickets_count": snapshot.period_created_tickets_count,
        "period_closed_tickets_count": snapshot.period_closed_tickets_count,
        "average_first_response_time_seconds": snapshot.average_first_response_time_seconds,
        "average_resolution_time_seconds": snapshot.average_resolution_time_seconds,
        "satisfaction_average": snapshot.satisfaction_average,
        "feedback_count": snapshot.feedback_count,
        "feedback_coverage_percent": snapshot.feedback_coverage_percent,
        "tickets_per_operator": [
            {
                "operator_id": item.operator_id,
                "display_name": item.display_name,
                "ticket_count": item.ticket_count,
            }
            for item in snapshot.tickets_per_operator
        ],
        "rating_distribution": [
            {
                "rating": bucket.rating,
                "count": bucket.count,
            }
            for bucket in snapshot.rating_distribution
        ],
        "operator_snapshots": [
            {
                "operator_id": item.operator_id,
                "display_name": item.display_name,
                "active_ticket_count": item.active_ticket_count,
                "closed_ticket_count": item.closed_ticket_count,
                "average_first_response_time_seconds": item.average_first_response_time_seconds,
                "average_resolution_time_seconds": item.average_resolution_time_seconds,
                "average_satisfaction": item.average_satisfaction,
                "feedback_count": item.feedback_count,
            }
            for item in snapshot.operator_snapshots
        ],
        "category_snapshots": [
            _serialize_category_snapshot(item) for item in snapshot.category_snapshots
        ],
        "best_operators_by_closures": [
            {
                "operator_id": item.operator_id,
                "display_name": item.display_name,
                "active_ticket_count": item.active_ticket_count,
                "closed_ticket_count": item.closed_ticket_count,
                "average_first_response_time_seconds": item.average_first_response_time_seconds,
                "average_resolution_time_seconds": item.average_resolution_time_seconds,
                "average_satisfaction": item.average_satisfaction,
                "feedback_count": item.feedback_count,
            }
            for item in snapshot.best_operators_by_closures
        ],
        "best_operators_by_satisfaction": [
            {
                "operator_id": item.operator_id,
                "display_name": item.display_name,
                "active_ticket_count": item.active_ticket_count,
                "closed_ticket_count": item.closed_ticket_count,
                "average_first_response_time_seconds": item.average_first_response_time_seconds,
                "average_resolution_time_seconds": item.average_resolution_time_seconds,
                "average_satisfaction": item.average_satisfaction,
                "feedback_count": item.feedback_count,
            }
            for item in snapshot.best_operators_by_satisfaction
        ],
        "top_categories": [_serialize_category_snapshot(item) for item in snapshot.top_categories],
        "first_response_breach_count": snapshot.first_response_breach_count,
        "resolution_breach_count": snapshot.resolution_breach_count,
        "sla_categories": [_serialize_category_snapshot(item) for item in snapshot.sla_categories],
    }


def _serialize_category_snapshot(item: Any) -> dict[str, Any]:
    return {
        "category_id": item.category_id,
        "category_title": item.category_title,
        "created_ticket_count": item.created_ticket_count,
        "open_ticket_count": item.open_ticket_count,
        "closed_ticket_count": item.closed_ticket_count,
        "average_satisfaction": item.average_satisfaction,
        "feedback_count": item.feedback_count,
        "sla_breach_count": item.sla_breach_count,
    }

File: src/mini_app/frontend/assets/renderers/analytics.js
import {
  renderBar,
  renderInlineEmpty,
  renderMetric,
} from "../render-utils.js";

export function renderAnalytics(data, windowKey) {
  const snapshot = data.snapshot;
  const windows = [
    ["today", "Сегодня"],
    ["7d", "7 дней"],
    ["30d", "30 дней"],
    ["all", "Всё время"],
  ];

  return `
    <section class="surface surface-roomy">
      <div class="surface-head">
        <div>
          <p class="eyebrow">Аналитика</p>
          <h2>Рабочая картина</h2>
          <p class="subtitle">Ключевые метрики по нагрузке, качеству и SLA.</p>
        </div>
        <div class="utility-actions utility-actions-end">
          <span class="utility-label">Выгрузка</span>
          <button class="action action-subtle" data-export-analytics="html">HTML</button>
          <button class="action action-subtle" data-export-analytics="csv">CSV</button>
        </div>
      </div>
      <div class="segmented-control">
        ${windows
          .map(
            ([value, label]) => `
              <button class="segmented ${windowKey === value ? "is-active" : ""}" data-window="${value}">
                ${label}
              </button>
            `,
          )
          .join("")}
      </div>
      <div class="metric-board">
        ${renderMetric("Открыто", snapshot.total_open_tickets)}
        ${renderMetric("Очередь", snapshot.queued_tickets_count)}
        ${renderMetric("Назначено", snapshot.assigned_tickets_count)}
        ${renderMetric("Эскалации", snapshot.escalated_tickets_count)}
        ${renderMetric("Средняя оценка", snapshot.satisfaction_average ? snapshot.satisfaction_average.toFixed(1) : "—")}
        ${renderMetric("Покрытие отзывов", snapshot.feedback_coverage_percent ? `${snapshot.feedback_coverage_percent}%` : "—")}
      </div>
      <div class="surface-grid">
        <article class="subsurface">
          <div class="subsurface-head">
            <h3>Темы</h3>
            <span class="soft-chip">По входящему потоку</span>
          </div>
          <div class="bars">
            ${snapshot.top_categories.length
              ? snapshot.top_categories
                  .slice(0, 6)
                  .map((item) =>
                    renderBar(
                      item.category_title,
                      item.created_ticket_count,
                      snapshot.period_created_tickets_count || 1,
                    ),
                  )
                  .join("")
              : renderInlineEmpty("Темы ещё не накопили историю.")}
          </div>
        </article>
        <article class="subsurface">
          <div class="subsurface-head">
            <h3>Команда</h3>
            <span class="soft-chip">По закрытиям</span>
          </div>
          <div class="bars">
            ${snapshot.operator_snapshots.length
              ? snapshot.operator_snapshots
                  .slice(0, 6)
                  .map((item) =>
                    renderBar(
                      item.display_name,
                      item.closed_ticket_count,
                      snapshot.period_closed_tickets_count || 1,
                      `${item.closed_ticket_count} закрыто`,
                    ),
                  )
                  .join("")
              : renderInlineEmpty("Статистика по команде появится после активности.")}
          </div>
        </article>
      </div>
      <div class="surface-grid">
        <article class="subsurface">
          <div class="subsurface-head">
            <h3>Распределение оценок</h3>
            <span class="soft-chip">Обратная связь</span>
          </div>
          <div class="bars">
            ${snapshot.rating_distribution.length
              ? snapshot.rating_distribution
                  .map((item) => renderBar(`${item.rating} / 5`, item.count, snapshot.feedback_count || 1))
                  .join("")
              : renderInlineEmpty("Оценок пока нет.")}
          </div>
        </article>
        <article class="subsurface">
          <div class="subsurface-head">
            <h3>SLA</h3>
            <span class="soft-chip">Контроль сроков</span>
          </div>
          <div class="stats-stack">
            ${renderMetric("Первый ответ", snapshot.first_response_breach_count)}
            ${renderMetric("Решение", snapshot.resolution_breach_count)}
          </div>
          <div class="bars">
            ${snapshot.sla_categories.length
              ? snapshot.sla_categories
                  .slice(0, 5)
                  .map((item) =>
                    renderBar(
                      item.category_title,
                      item.sla_breach_count,
                      Math.max(snapshot.first_response_breach_count + snapshot.resolution_breach_count, 1),
                    ),
                  )
                  .join("")
              : renderInlineEmpty("Нарушений SLA не зафиксировано.")}
          </div>
        </article>
      </div>
    </section>
  `;
}
